File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-option-shared-5-4/Knock.cpp
// Knock.cpp 
#include "Knock.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

Knock::Knock(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //knockInNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* knockInNode = xmlNode->FirstChildElement("knockIn");

   if(knockInNode){knockInIsNull_ = false;}
   else{knockInIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- knockInNode , address : " << knockInNode << std::endl;
   #endif
   if(knockInNode)
   {
      if(knockInNode->Attribute("href") || knockInNode->Attribute("id"))
      {
          if(knockInNode->Attribute("id"))
          {
             knockInIDRef_ = knockInNode->Attribute("id");
             knockIn_ = boost::shared_ptr<TriggerEvent>(new TriggerEvent(knockInNode));
             knockIn_->setID(knockInIDRef_);
             IDManager::instance().setID(knockInIDRef_,knockIn_);
          }
          else if(knockInNode->Attribute("href")) { knockInIDRef_ = knockInNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { knockIn_ = boost::shared_ptr<TriggerEvent>(new TriggerEvent(knockInNode));}
   }

   //knockOutNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* knockOutNode = xmlNode->FirstChildElement("knockOut");

   if(knockOutNode){knockOutIsNull_ = false;}
   else{knockOutIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- knockOutNode , address : " << knockOutNode << std::endl;
   #endif
   if(knockOutNode)
   {
      if(knockOutNode->Attribute("href") || knockOutNode->Attribute("id"))
      {
          if(knockOutNode->Attribute("id"))
          {
             knockOutIDRef_ = knockOutNode->Attribute("id");
             knockOut_ = boost::shared_ptr<TriggerEvent>(new TriggerEvent(knockOutNode));
             knockOut_->setID(knockOutIDRef_);
             IDManager::instance().setID(knockOutIDRef_,knockOut_);
          }
          else if(knockOutNode->Attribute("href")) { knockOutIDRef_ = knockOutNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { knockOut_ = boost::shared_ptr<TriggerEvent>(new TriggerEvent(knockOutNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<TriggerEvent> Knock::getKnockIn()
{
   if(!this->knockInIsNull_){
        if(knockInIDRef_ != ""){
             return boost::shared_static_cast<TriggerEvent>(IDManager::instance().getID(knockInIDRef_));
        }else{
             return this->knockIn_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<TriggerEvent>();
   }
}
boost::shared_ptr<TriggerEvent> Knock::getKnockOut()
{
   if(!this->knockOutIsNull_){
        if(knockOutIDRef_ != ""){
             return boost::shared_static_cast<TriggerEvent>(IDManager::instance().getID(knockOutIDRef_));
        }else{
             return this->knockOut_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<TriggerEvent>();
   }
}
}

File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-option-shared-5-4/Knock.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class Knock : ISerialized
    {
        public Knock(XmlNode xmlNode)
        {
            XmlNode knockInNode = xmlNode.SelectSingleNode("knockIn");
            
            if (knockInNode != null)
            {
                if (knockInNode.Attributes["href"] != null || knockInNode.Attributes["id"] != null) 
                {
                    if (knockInNode.Attributes["id"] != null) 
                    {
                        knockInIDRef_ = knockInNode.Attributes["id"].Value;
                        TriggerEvent ob = new TriggerEvent(knockInNode);
                        IDManager.SetID(knockInIDRef_, ob);
                    }
                    else if (knockInNode.Attributes["href"] != null)
                    {
                        knockInIDRef_ = knockInNode.Attributes["href"].Value;
                    }
                    else
                    {
                        knockIn_ = new TriggerEvent(knockInNode);
                    }
                }
                else
                {
                    knockIn_ = new TriggerEvent(knockInNode);
                }
            }
            
        
            XmlNode knockOutNode = xmlNode.SelectSingleNode("knockOut");
            
            if (knockOutNode != null)
            {
                if (knockOutNode.Attributes["href"] != null || knockOutNode.Attributes["id"] != null) 
                {
                    if (knockOutNode.Attributes["id"] != null) 
                    {
                        knockOutIDRef_ = knockOutNode.Attributes["id"].Value;
                        TriggerEvent ob = new TriggerEvent(knockOutNode);
                        IDManager.SetID(knockOutIDRef_, ob);
                    }
                    else if (knockOutNode.Attributes["href"] != null)
                    {
                        knockOutIDRef_ = knockOutNode.Attributes["href"].Value;
                    }
                    else
                    {
                        knockOut_ = new TriggerEvent(knockOutNode);
                    }
                }
                else
                {
                    knockOut_ = new TriggerEvent(knockOutNode);
                }
            }
            
        
        }
        
    
        #region KnockIn_
        private TriggerEvent knockIn_;
        public TriggerEvent KnockIn_
        {
            get
            {
                if (this.knockIn_ != null)
                {
                    return this.knockIn_; 
                }
                else if (this.knockInIDRef_ != null)
                {
                    knockIn_ = IDManager.getID(knockInIDRef_) as TriggerEvent;
                    return this.knockIn_; 
                }
                else
                {
                      return this.knockIn_; 
                }
            }
            set
            {
                if (this.knockIn_ != value)
                {
                    this.knockIn_ = value;
                }
            }
        }
        #endregion
        
        public string knockInIDRef_ { get; set; }
        #region KnockOut_
        private TriggerEvent knockOut_;
        public TriggerEvent KnockOut_
        {
            get
            {
                if (this.knockOut_ != null)
                {
                    return this.knockOut_; 
                }
                else if (this.knockOutIDRef_ != null)
                {
                    knockOut_ = IDManager.getID(knockOutIDRef_) as TriggerEvent;
                    return this.knockOut_; 
                }
                else
                {
                      return this.knockOut_; 
                }
            }
            set
            {
                if (this.knockOut_ != value)
                {
                    this.knockOut_ = value;
                }
            }
        }
        #endregion
        
        public string knockOutIDRef_ { get; set; }
        
    
        
    
    }
    
}

